src/atlas_f/util/atlas_Geometry_module.F90

```diff
@@ -0,0 +1,179 @@
+
+interface atlas_Geometry
+  module procedure atlas_Geometry__ctor_name
+  module procedure atlas_Geometry__ctor_radius
+end interface
+
+!------------------------------------------------------------------------------
+
+
+!========================================================
+contains
+!========================================================
+
+
+! -----------------------------------------------------------------------------
+! Geometry routines
+
+function atlas_Geometry__ctor_name(name) result(this)
+  use atlas_Geometry_c_binding
+  use fckit_c_interop_module
+  character(kind=c_char,len=*), intent(in) :: name
+  type(atlas_Geometry) :: this
+  call this%reset_c_ptr( atlas__Geometry__new_name( c_str(name) ) )
+  call this%return()
+end function atlas_Geometry__ctor_name
+
+function atlas_Geometry__ctor_radius(radius) result(this)
+  use atlas_Geometry_c_binding
+  use fckit_c_interop_module
+  real(c_double), intent(in) :: radius
+  type(atlas_Geometry) :: this
+  call this%reset_c_ptr( atlas__Geometry__new_radius( radius ) )
+  call this%return()
+end function atlas_Geometry__ctor_radius
+
+subroutine atlas_Geometry__delete(this)
+  use atlas_Geometry_c_binding
+  class(atlas_Geometry), intent(inout) :: this
+  if ( .not. this%is_null() ) then
+    call atlas__Geometry__delete(this%CPTR_PGIBUG_A)
+  end if
+  call this%reset_c_ptr()
+end subroutine atlas_Geometry__delete
+
+subroutine Geometry__xyz2lonlat(this, x, y, z, lon, lat)
+  use atlas_Geometry_c_binding
+  use fckit_c_interop_module
+  class(atlas_Geometry), intent(in) :: this
+  real(c_double), intent(in) :: x
+  real(c_double), intent(in) :: y
+  real(c_double), intent(in) :: z
+  real(c_double), intent(out) :: lon
+  real(c_double), intent(out) :: lat
+  call atlas__Geometry__xyz2lonlat(this%CPTR_PGIBUG_A, x, y, z, lon, lat)
+end subroutine Geometry__xyz2lonlat
+
+subroutine Geometry__lonlat2xyz(this, lon, lat, x, y, z)
+  use atlas_Geometry_c_binding
+  use fckit_c_interop_module
+  class(atlas_Geometry), intent(in) :: this
+  real(c_double), intent(in) :: lon
+  real(c_double), intent(in) :: lat
+  real(c_double), intent(out) :: x
+  real(c_double), intent(out) :: y
+  real(c_double), intent(out) :: z
+  call atlas__Geometry__lonlat2xyz(this%CPTR_PGIBUG_A, lon, lat, x, y, z)
+end subroutine Geometry__lonlat2xyz
+
+function Geometry__distance_lonlat(this, lon1, lat1, lon2, lat2) result(distance_lonlat)
+  use atlas_Geometry_c_binding
+  use fckit_c_interop_module
+  class(atlas_Geometry), intent(in) :: this
+  real(c_double), intent(in) :: lon1
+  real(c_double), intent(in) :: lat1
+  real(c_double), intent(in) :: lon2
+  real(c_double), intent(in) :: lat2
+  real(c_double) :: distance_lonlat
+  distance_lonlat = atlas__Geometry__distance_lonlat(this%CPTR_PGIBUG_A, lon1, lat1, lon2, lat2)
+end function Geometry__distance_lonlat
+
+function Geometry__distance_xyz(this, x1, y1, z1, x2, y2, z2) result(distance_xyz)
```

Comment:
Can you create another overload `distance_xyz( p1, p2 )
where p1 and p2 are size(3) arrays?